File: apps/android/app/src/main/java/com/michaeldadi/sprout/services/BiometricAuthService.kt
package com.michaeldadi.sprout.services

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume

sealed class BiometricAuthError(val message: String) {
    object BiometryNotAvailable : BiometricAuthError("Biometric authentication is not available on this device")
    object BiometryNotEnrolled : BiometricAuthError("No biometric data is enrolled on this device")
    object BiometryLockout : BiometricAuthError("Biometric authentication is locked out. Try again later")
    object AuthenticationFailed : BiometricAuthError("Authentication failed")
    object UserCancel : BiometricAuthError("Authentication was cancelled by user")
    object SystemCancel : BiometricAuthError("Authentication was cancelled by system")
    object PasscodeNotSet : BiometricAuthError("Device credential is not set on this device")
    class Unknown(error: String) : BiometricAuthError("Unknown error: $error")
}

enum class BiometricType(val displayName: String, val iconResource: String?) {
    NONE("None", null),
    FINGERPRINT("Fingerprint", "ic_fingerprint"),
    FACE("Face", "ic_face"),
    IRIS("Iris", "ic_iris");
}

class BiometricAuthService private constructor(private val context: Context) {
    companion object {
        @Volatile
        private var INSTANCE: BiometricAuthService? = null

        fun getInstance(context: Context): BiometricAuthService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: BiometricAuthService(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences("biometric_auth_prefs", Context.MODE_PRIVATE)
    private val biometricEnabledKey = "biometric_auth_enabled"

    var isEnabled: Boolean
        get() = sharedPreferences.getBoolean(biometricEnabledKey, false)
        set(value) {
            sharedPreferences.edit().putBoolean(biometricEnabledKey, value).apply()
        }

    val isAvailable: Boolean
      get() = checkBiometricAvailability() == BiometricManager.BIOMETRIC_SUCCESS

    val biometricType: BiometricType
        get() = when (checkBiometricAvailability()) {
            BiometricManager.BIOMETRIC_SUCCESS -> {
                // Android doesn't provide specific biometric type info easily
                // We'll default to fingerprint for available biometrics
                BiometricType.FINGERPRINT
            }
            else -> BiometricType.NONE
        }

    private fun checkBiometricAvailability(): Int {
        val biometricManager = BiometricManager.from(context)
        return biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)
    }

    suspend fun authenticate(
        activity: FragmentActivity,
        title: String = "Biometric Authentication",
        subtitle: String = "Please authenticate to access your account",
        description: String = "Use your biometric credential to authenticate"
    ): Result<Boolean> = suspendCancellableCoroutine { continuation ->

        if (!isAvailable) {
            continuation.resume(Result.failure(Exception(BiometricAuthError.BiometryNotAvailable.message)))
            return@suspendCancellableCoroutine
        }

        if (!isEnabled) {
            continuation.resume(Result.failure(Exception(BiometricAuthError.BiometryNotAvailable.message)))
            return@suspendCancellableCoroutine
        }

        val executor = ContextCompat.getMainExecutor(context)
        val biometricPrompt = BiometricPrompt(activity, executor, object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                super.onAuthenticationError(errorCode, errString)
                val error = when (errorCode) {
                    BiometricPrompt.ERROR_CANCELED -> BiometricAuthError.SystemCancel
                    BiometricPrompt.ERROR_USER_CANCELED -> BiometricAuthError.UserCancel
                    BiometricPrompt.ERROR_LOCKOUT, BiometricPrompt.ERROR_LOCKOUT_PERMANENT -> BiometricAuthError.BiometryLockout
                    BiometricPrompt.ERROR_NO_BIOMETRICS -> BiometricAuthError.BiometryNotEnrolled
                    BiometricPrompt.ERROR_HW_NOT_PRESENT, BiometricPrompt.ERROR_HW_UNAVAILABLE -> BiometricAuthError.BiometryNotAvailable
                    BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL -> BiometricAuthError.PasscodeNotSet
                    else -> BiometricAuthError.Unknown(errString.toString())
                }
                Log.e("BiometricAuthService", "Authentication error: $errString")
                if (continuation.isActive) {
                    continuation.resume(Result.failure(Exception(error.message)))
                }
            }

            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                super.onAuthenticationSucceeded(result)
                Log.i("BiometricAuthService", "Authentication succeeded")
                if (continuation.isActive) {
                    continuation.resume(Result.success(true))
                }
            }

            override fun onAuthenticationFailed() {
                super.onAuthenticationFailed()
                Log.w("BiometricAuthService", "Authentication failed")
                if (continuation.isActive) {
                    continuation.resume(Result.failure(Exception(BiometricAuthError.AuthenticationFailed.message)))
                }
            }
        })

        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setDescription(description)
            .setNegativeButtonText("Cancel")
            .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_WEAK)
            .build()

        activity.lifecycleScope.launch {
          activity.lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
            biometricPrompt.authenticate(promptInfo)
          }
        }

        continuation.invokeOnCancellation {
            biometricPrompt.cancelAuthentication()
        }
    }

    suspend fun authenticateWithPasscode(
        activity: FragmentActivity,
        title: String = "Device Authentication",
        subtitle: String = "Please authenticate to access your account",
        description: String = "Use your device PIN, pattern, or password to authenticate"
    ): Result<Boolean> = suspendCancellableCoroutine { continuation ->

        val executor = ContextCompat.getMainExecutor(context)
        val biometricPrompt = BiometricPrompt(activity, executor, object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                super.onAuthenticationError(errorCode, errString)
                val error = when (errorCode) {
                    BiometricPrompt.ERROR_CANCELED -> BiometricAuthError.SystemCancel
                    BiometricPrompt.ERROR_USER_CANCELED -> BiometricAuthError.UserCancel
                    BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL -> BiometricAuthError.PasscodeNotSet
                    else -> BiometricAuthError.Unknown(errString.toString())
                }
                Log.e("BiometricAuthService", "Passcode authentication error: $errString")
                if (continuation.isActive) {
                    continuation.resume(Result.failure(Exception(error.message)))
                }
            }

            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                super.onAuthenticationSucceeded(result)
                Log.i("BiometricAuthService", "Passcode authentication succeeded")
                if (continuation.isActive) {
                    continuation.resume(Result.success(true))
                }
            }

            override fun onAuthenticationFailed() {
                super.onAuthenticationFailed()
                Log.w("BiometricAuthService", "Passcode authentication failed")
                if (continuation.isActive) {
                    continuation.resume(Result.failure(Exception(BiometricAuthError.AuthenticationFailed.message)))
                }
            }
        })

        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setDescription(description)
            .setAllowedAuthenticators(
                BiometricManager.Authenticators.BIOMETRIC_WEAK or
                BiometricManager.Authenticators.DEVICE_CREDENTIAL
            )
            .build()

        activity.lifecycleScope.launch {
          activity.lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
            biometricPrompt.authenticate(promptInfo)
          }
        }

        continuation.invokeOnCancellation {
            biometricPrompt.cancelAuthentication()
        }
    }

    fun enableBiometricAuth() {
        isEnabled = true
        Log.i("BiometricAuthService", "Biometric authentication enabled")
    }

    fun disableBiometricAuth() {
        isEnabled = false
        Log.i("BiometricAuthService", "Biometric authentication disabled")
    }

    fun getBiometricCapabilityString(): String {
        return when (checkBiometricAvailability()) {
            BiometricManager.BIOMETRIC_SUCCESS -> "Biometric authentication available"
            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE -> "No biometric hardware available"
            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> "Biometric hardware unavailable"
            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> "No biometric credentials enrolled"
            BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED -> "Security update required"
            BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED -> "Biometric authentication unsupported"
            BiometricManager.BIOMETRIC_STATUS_UNKNOWN -> "Biometric status unknown"
            else -> "Unknown biometric status"
        }
    }
}

File: apps/ios/Sprout/Services/BiometricAuthService.swift
import Foundation
import LocalAuthentication
import SwiftUI

enum BiometricAuthError: Error {
    case biometryNotAvailable
    case biometryNotEnrolled
    case biometryLockout
    case authenticationFailed
    case userCancel
    case userFallback
    case systemCancel
    case passcodeNotSet
    case unknown(Error)
    
    var localizedDescription: String {
        switch self {
        case .biometryNotAvailable:
            return "Biometric authentication is not available on this device"
        case .biometryNotEnrolled:
            return "No biometric data is enrolled on this device"
        case .biometryLockout:
            return "Biometric authentication is locked out. Use passcode instead"
        case .authenticationFailed:
            return "Authentication failed"
        case .userCancel:
            return "Authentication was cancelled by user"
        case .userFallback:
            return "User chose to use passcode instead"
        case .systemCancel:
            return "Authentication was cancelled by system"
        case .passcodeNotSet:
            return "Passcode is not set on this device"
        case .unknown(let error):
            return "Unknown error: \(error.localizedDescription)"
        }
    }
}

enum BiometricType {
    case none
    case touchID
    case faceID
    case opticID
    
    var displayName: String {
        switch self {
        case .none:
            return "None"
        case .touchID:
            return "Touch ID"
        case .faceID:
            return "Face ID"
        case .opticID:
            return "Optic ID"
        }
    }
    
    var iconName: String {
        switch self {
        case .none:
            return "lock"
        case .touchID:
            return "touchid"
        case .faceID:
            return "faceid"
        case .opticID:
            return "opticid"
        }
    }
}

@MainActor
class BiometricAuthService: ObservableObject {
    static let shared = BiometricAuthService()
    
    @Published var isEnabled: Bool = false
    @Published var biometricType: BiometricType = .none
    @Published var isAvailable: Bool = false
    
    private let context = LAContext()
    private let userDefaults = UserDefaults.standard
    private let biometricEnabledKey = "biometric_auth_enabled"
    
    private init() {
        loadSettings()
        checkBiometricAvailability()
    }
    
    func loadSettings() {
        isEnabled = userDefaults.bool(forKey: biometricEnabledKey)
    }
    
    func saveSettings() {
        userDefaults.set(isEnabled, forKey: biometricEnabledKey)
    }
    
    func checkBiometricAvailability() {
        var error: NSError?
        let isAvailableResult = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
        
        isAvailable = isAvailableResult
        
        if isAvailableResult {
            switch context.biometryType {
            case .touchID:
                biometricType = .touchID
            case .faceID:
                biometricType = .faceID
            case .opticID:
                biometricType = .opticID
            case .none:
                biometricType = .none
                isAvailable = false
            @unknown default:
                biometricType = .none
                isAvailable = false
            }
        } else {
            biometricType = .none
        }
    }
    
    func authenticate(reason: String = "Please authenticate to access your account") async -> Result<Bool, BiometricAuthError> {
        guard isAvailable else {
            return .failure(.biometryNotAvailable)
        }
        
        guard isEnabled else {
            return .failure(.biometryNotAvailable)
        }
        
        let context = LAContext()
        context.localizedFallbackTitle = "Use Passcode"
        context.localizedCancelTitle = "Cancel"
        
        do {
            let result = try await context.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: reason
            )
            return .success(result)
        } catch {
            let authError = mapLAError(error)
            return .failure(authError)
        }
    }
    
    func authenticateWithPasscode(reason: String = "Please authenticate to access your account") async -> Result<Bool, BiometricAuthError> {
        let context = LAContext()
        context.localizedFallbackTitle = ""
        context.localizedCancelTitle = "Cancel"
        
        do {
            let result = try await context.evaluatePolicy(
                .deviceOwnerAuthentication,
                localizedReason: reason
            )
            return .success(result)
        } catch {
            let authError = mapLAError(error)
            return .failure(authError)
        }
    }
    
    func enableBiometricAuth() {
        isEnabled = true
        saveSettings()
    }
    
    func disableBiometricAuth() {
        isEnabled = false
        saveSettings()
    }
    
    private func mapLAError(_ error: Error) -> BiometricAuthError {
        guard let laError = error as? LAError else {
            return .unknown(error)
        }
        
        switch laError.code {
        case .biometryNotAvailable:
            return .biometryNotAvailable
        case .biometryNotEnrolled:
            return .biometryNotEnrolled
        case .biometryLockout:
            return .biometryLockout
        case .authenticationFailed:
            return .authenticationFailed
        case .userCancel:
            return .userCancel
        case .userFallback:
            return .userFallback
        case .systemCancel:
            return .systemCancel
        case .passcodeNotSet:
            return .passcodeNotSet
        default:
            return .unknown(error)
        }
    }
}